### PR TITLE
Implement encoder for AttestationData

### PR DESCRIPTION
**Pull request by awly**
Also export Encode methods for Public and Private structs. These allow
testing of attestation verification code by mocking encoded
AttestationData.

### CODE REVIEW

File: tpm2/structures.go [hunk 1]
```diff
@@ -469,6 +495,18 @@ func decodeCertifyInfo(in *bytes.Buffer) (*CertifyInfo, error) {
 	return &ci, nil
 }
 
+func (ci CertifyInfo) encode() ([]byte, error) {
+	n, err := ci.Name.encode()
+	if err != nil {
+		return nil, fmt.Errorf("encoding Name: %v", err)
```
[the remark this answers is not in the document]
Comment (awly) [reply tpm2/structures.go]: I think all other encoding functions could use more detail too. When encoding error happens on some deeply-nested field, this info helps pinpoint the exact value that failed to encode.

Happy to fix all other errors in followup PR or in this one.

File: tpm2/constants.go [hunk 2]
```diff
@@ -148,12 +148,13 @@ const (
 	CapabilityAuthPolicies
 )
 
+// TPM Tags.
```
[the remark this answers is not in the document]
Comment (awly) [reply tpm2/constants.go]: Done

File: tpm2/structures.go [hunk 3]
```diff
@@ -64,7 +64,8 @@ type Public struct {
 	ECCParameters *ECCParams
 }
 
-func (p Public) encode() ([]byte, error) {
+// Encode returns Public serialized to TPM wire format.
```
[the remark this answers is not in the document]
Comment (awly) [reply tpm2/structures.go]: Done

File: tpm2/structures.go [hunk 4]
```diff
@@ -501,10 +539,30 @@ func decodeName(in *bytes.Buffer) (*Name, error) {
 	return name, nil
 }
 
+func (n Name) encode() ([]byte, error) {
+	var buf []byte
+	var err error
+	switch {
+	case n.Handle != nil:
+		buf, err = tpmutil.Pack(*n.Handle)
```
[the remark this answers is not in the document]
Comment (awly) [reply tpm2/structures.go]: Done

File: tpm2/tpm2_test.go [hunk 5]
```diff
@@ -573,3 +573,47 @@ func TestReadPCR(t *testing.T) {
...
+			},
+			QualifiedName: Name{
+				Handle: &ciQualifiedName,
+			},
+		},
+	}
+
+	encoded, err := ad.Encode()
+	if err != nil {
+		t.Fatalf("error encoding: %v", err)
```
[the remark this answers is not in the document]
Comment (awly) [reply tpm2/tpm2_test.go]: Done

File: tpm2/tpm2_test.go [hunk 6]
```diff
@@ -573,3 +573,47 @@ func TestReadPCR(t *testing.T) {
...
+	}
+	decoded, err := DecodeAttestationData(encoded)
+	if err != nil {
+		t.Fatalf("error decoding: %v", err)
+	}
+
+	if !reflect.DeepEqual(*decoded, ad) {
+		t.Errorf("got decoded value:\n%#v\nwant:\n%#v", decoded, ad)
```
[the remark this answers is not in the document]
Comment (awly) [reply tpm2/tpm2_test.go]: If the test fails, the extra info from `%#v` (field names and types) helps with debugging.
But it's not really necessary, changed to `%v`

File: tpm2/structures.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tmroeder) [reply tpm2/structures.go]: I think it's fine to do this in a followup, since that's not the purpose of this PR.

**Pull request merged**